remote/remote_state_gcs.go

```diff
@@ -0,0 +1,405 @@
+	// A project must be specified in order for terragrunt to automatically create a storage bucket.
+	if extendedConfig.Project == "" {
+		return errors.WithStackTrace(MissingRequiredGCSRemoteStateConfig("project"))
+	}
+
+	var config = extendedConfig.remoteStateConfigGCS
+
+	if config.Bucket == "" {
+		return errors.WithStackTrace(MissingRequiredGCSRemoteStateConfig("bucket"))
+	}
+
+	if config.Prefix == "" {
+		return errors.WithStackTrace(MissingRequiredGCSRemoteStateConfig("prefix"))
+	}
+
+	return nil
+}
+
+// If the bucket specified in the given config doesn't already exist, prompt the user to create it, and if the user
+// confirms, create the bucket and enable versioning for it.
+func createGCSBucketIfNecessary(gcsClient *storage.Client, config *ExtendedRemoteStateConfigGCS, terragruntOptions *options.TerragruntOptions) error {
+	if !DoesGCSBucketExist(gcsClient, &config.remoteStateConfigGCS) {
+		prompt := fmt.Sprintf("Remote state GCS bucket %s does not exist or you don't have permissions to access it. Would you like Terragrunt to create it?", config.remoteStateConfigGCS.Bucket)
+		shouldCreateBucket, err := shell.PromptUserForYesNo(prompt, terragruntOptions)
+		if err != nil {
+			return err
+		}
+
+		if shouldCreateBucket {
+			// To avoid any eventual consistency issues with creating a GCS bucket we use a retry loop.
+			description := fmt.Sprintf("Create GCS bucket %s", config.remoteStateConfigGCS.Bucket)
+			maxRetries := 3
+			sleepBetweenRetries := 10 * time.Second
+
+			return util.DoWithRetry(description, maxRetries, sleepBetweenRetries, terragruntOptions.Logger, func() error {
+				return CreateGCSBucketWithVersioning(gcsClient, config, terragruntOptions)
+			})
+		}
+	}
+
+	return nil
+}
+
+// Check if versioning is enabled for the GCS bucket specified in the given config and warn the user if it is not
+func checkIfGCSVersioningEnabled(gcsClient *storage.Client, config *RemoteStateConfigGCS, terragruntOptions *options.TerragruntOptions) error {
+	ctx := context.Background()
+	bucket := gcsClient.Bucket(config.Bucket)
+
+	attrs, err := bucket.Attrs(ctx)
+	if err != nil {
+		// ErrBucketNotExist
+		return errors.WithStackTrace(err)
+	}
+
+	if attrs.VersioningEnabled == false {
+		terragruntOptions.Logger.Printf("WARNING: Versioning is not enabled for the remote state GCS bucket %s. We recommend enabling versioning so that you can roll back to previous versions of your Terraform state in case of error.", config.Bucket)
+	}
+
+	return nil
+}
+
+// CreateGCSBucketWithVersioning creates the given GCS bucket and enables versioning for it.
+func CreateGCSBucketWithVersioning(gcsClient *storage.Client, config *ExtendedRemoteStateConfigGCS, terragruntOptions *options.TerragruntOptions) error {
+	err := CreateGCSBucket(gcsClient, config, terragruntOptions)
+
+	if err != nil {
+		if isBucketAlreadyOwnedByYourError(err) {
```

Comment:
Is this using the same method from the S3 code? If so, will that work for GCS?

Reply:
No, I'll remove it.